include/dpp/coro/task.h

```diff
@@ -375,136 +134,56 @@ class task : private detail::task::task_base<R> {
 	 *
 	 * @param other Task to move the handle from
 	 */
-	task &operator=(task &&other) noexcept;
-
-	/**
-	 * @brief Function to check if the task has finished its execution entirely
-	 *
-	 * @return bool Whether the task is finished.
-	 */
-	[[nodiscard]] bool done() const noexcept;
-
-	/**
-	 * @brief Cancel the task, it will stop the next time it uses co_await. On co_await-ing this task, throws dpp::task_cancelled_exception.
-	 */
-	dpp::task<R>& cancel() & noexcept;
-
-	/**
-	 * @brief Check whether or not a call to co_await will suspend the caller.
-	 *
-	 * This function is called by the standard library as a first step when using co_await. If it returns true then the caller is not suspended.
-	 * @throws logic_exception if the task is empty.
-	 * @return bool Whether not to suspend the caller or not
-	 */
-	[[nodiscard]] bool await_ready() const;
-#else
-	using detail::task::task_base<R>::task_base; // use task_base's constructors
-	using detail::task::task_base<R>::operator=; // use task_base's assignment operators
-	using detail::task::task_base<R>::done; // expose done() as public
-	using detail::task::task_base<R>::cancel; // expose cancel() as public
-	using detail::task::task_base<R>::await_ready; // expose await_ready as public
-#endif
-
-	/**
-	 * @brief Suspend the current coroutine until the task completes.
-	 *
-	 * @throw On resumption, any exception thrown by the coroutine is propagated to the caller.
-	 * @return On resumption, this expression evaluates to the result object of type R, as a reference.
-	 */
-	[[nodiscard]] auto& operator co_await() & noexcept {
-		return static_cast<detail::task::task_base<R>&>(*this);
+	task &operator=(task &&other) noexcept {
+		awaitable<R>::operator=(std::move(other));
+		handle = std::exchange(other.handle, nullptr);
+		return *this;
 	}
 
 	/**
-	 * @brief Suspend the current coroutine until the task completes.
+	 * @brief Destructor.
 	 *
-	 * @throw On resumption, any exception thrown by the coroutine is propagated to the caller.
-	 * @return On resumption, this expression evaluates to the result object of type R, as a const reference.
+	 * Destroys the handle. If the task is still running, it will be cancelled.
 	 */
-	[[nodiscard]] const auto& operator co_await() const & noexcept {
-		return static_cast<const detail::task::task_base<R>&>(*this);
+	~task() {
+		if (handle && this->valid()) {
+			if (this->abandon() & state_flags::sf_done) {
```

Comment:
note to self: don't use an implicit conversion to bool

Reply:
note to self: note